cancelable/closer.go

```diff
@@ -12,24 +12,65 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+// Package cancelable provides a cancelable version of io.Closer.
+//
+// This is intended to be used by functions where the responsibility for closing
+// a resource sometimes belongs to the function itself (e.g. early returns due
+// to error handling), and sometimes belongs to the caller.
+//
+// In the example below, this responsibility is shared by both f() and caller().
+// There is always a deferred function that is ready to close the connection
+// when needed, and we never need to worry about forgetting to call Close().
+//
+//  func f() (net.Conn, error) {
```

Comment:
You could put this in `example_closer_test.go` so it can function as a test and show as an example in the docs. I thought of this as I started reviewing the code in the comment, wondering if there were any typos. Although I admit a real example might be hard in the playground environment that examples run in.

If you haven't looked at testable examples before, start here: https://blog.golang.org/examples

You can see what we have done at https://pkg.go.dev/foxygo.at/s/httpe (example source at https://github.com/foxygoat/s/tree/master/httpe)

Reply:
Yeah I was thinking of doing an example, but then I realized that the "play" button wouldn't work due to the sandbox not allowing any network requests. Happy to put up an example if you can think of another Closer that could be wrapped, and works in the sandbox...

Reply:
You could just knock up an example closer that prints "Closed" (or not, if cancelled). It would only be a couple of extra lines.

Reply:
Makes sense, I've just done that. But it doesn't seem to work when I add `//output:` comments, or when I try to have multiple examples (possibly because I've got a too-old version of `godoc`?). Hopefully the example isn't too noisy, let me know what you think?